### ISSUE TITLE
QNRG Classifier ML Model Improvements

### ISSUE THREAD

**Issue by ShreyK7** [dorahacksglobal/qc-classifier#3]
We are building a classifier/prediction model that attempts to classify quantum computers based on quantum random number (QNRG) output data. We are looking to improve the performance (in other words, prediction accuracy) of our classifier model. We know that if we concatenate our binary QNRG sequences to increase the size of each training data line, the model is able to reach increasing accuracy, so we believe that it is possible for a model to distinguish between quantum computers with high accuracy - specifically because each QC has its unique quantum noise.

We are currently using an optimized gradient booster model as our main prediction model to train and test our QNRG input data (included in the repo). We have found that layering ('stacking') decision trees is currently how we can maximize the prediction accuracy when our model is applied to the processed (so after all preprocessing features are added, in other words) QNRG training data. However, given the vast landscape of prediction and modeling techniques that exist in today's ML/AI landscape, as well as the uniqueness of this quantum computing problem, there are almost certainly ways in which our model can be altered/changed to improve our current QC classification accuracy. Whether by repurposing or altering the gradient booster model in the repo notebook or creating your own prediction model (such as exploring different types of neural network techniques, for example), we would love to see if you can significantly improve the current prediction accuracy of our end-to-end classifier model. 

The benchmark and current model accuracy (with the optimized combination of features we have already tested before training our gradient booster model) is described in this repo's readme, so any improvements (to that score) will be useful and in the scope of the bounty! Please feel free to comment on this issue or email [shrey@dorahacks.biz](mailto:shrey@dorahacks.biz) & [sid@dorahacks.biz](mailto:sid@dorahacks.biz) if you have any questions.

You can duplicate the main jupyter notebook in this repo to play around with the models we have build and tested, and add your own methods. Good luck!

**Comment by neogyk** [comment]
Dear @ShreyK7 , I would like to improve the classifier.

**Comment by neogyk** [comment]
I have the model that predict the test dataset with 
Accuracy:  0.5408333333333334

**Comment by ShreyK7** [comment]
Apologies, we just realized that the data file in the repo was a test data file with only 99 lines of QNRG output data. We have replaced that file with the full 14000 QNRG data file in the repo so you can test your models more robustly. Again, sorry for the confusion, the full-length data file is now in this repo.

**Comment by ShreyK7** [comment]
> > The labels at the end of each line classify which quantum computer the data was generated from (there are 4 devices we used).
> 
> Can you share which 4 devices these were? It would help to assign meaning to the labels, and I didn't see this documented in the other experiment repositories.

Hi @schance995 Unfortunately not sure which specific devices the data was generated from, but I believe they are all AWS connected quantum computers. Label 4 is Rigetti's Aspen-M-3 quantum computer, but not sure about the others. @bernardyeszero might have more information

**Comment by ShreyK7** [comment]
@idriss-hamadi yes that is excellent if done properly! Submit a pull request with your new/changed code or jupyter notebook and we will take a look.

**Comment by idriss-hamadi** [comment]
please update me when you test the notebook

i plan to tackle the pre-processing issue this Wednesday

**Comment by ShreyK7** [comment]
@idriss-hamadi - @sid-chava left a review on your PR

**Comment by neogyk** [comment]
Hey @ShreyK7, 
I retrained the model using new dataset using the raw data, without processing it.
Here is the results:

Early stopping occured at epoch 1526 with best_epoch = 526 and best_val_1_accuracy = 0.58143
Best weights from best epoch are automatically used!
Accuracy:  0.5814285714285714
Confusion matrix:
[[ 103   84   44  172]
 [ 118  189    9  122]
 [  64   29  230  193]
 [ 122   96  119 1106]]

**Comment by sid-chava** [comment]
Nice. I would say if you are interested in the preprocessing bounty I would recommend moving to that at this point - can you also submit/update the pull request with the changes?

**Comment by neogyk** [comment]
@sid-chava Yes, I can upload the notebook

**Comment by ShreyK7** [comment]
@shubhusion Very nice - submit a pr so we can review your model (you can make a new Jupyter notebook with your updated code).

**Comment by idriss-hamadi** [comment]
@ShreyK7 i did update my PR, and explained the new approach more in the review section, please update me when you check it and have a review about it

**Comment by idriss-hamadi** [comment]
another question, will we be having more data to work on ?

**Comment by ShreyK7** [comment]
@idriss-hamadi Sounds good! Will take a look later today or early tomorrow. As for your second question, we are working to generate more data through AWS/Azure/QBraid, but unfortunately most QPUs such as provided by IonQ are currently down and/or not currently accessible. We will try to generate more training data by Monday if possible.
